Initial Commit of CLI

odgrace opened the pull request:
First shot at making a CLI utility to allow batch processing

odgrace committed:
Initial Commit of CLI

kirek007 commented:
Hey @odgrace this is exactly how I would do this! I'll take time later today to review and suggest few improvements before merge.

Thanks for contributing!

kirek007 commented on cli.py:
```diff
@@ -0,0 +1,72 @@
+        osd_path=implicit_path(args, 'osd'),
+        srt_path=implicit_path(args, 'srt'),
+        font_path=args.font_path,
+        output_path='/tmp/files_out3', # TODO make this video only and png automatic
```
Let's make this an input parameter with default value of current directory

odgrace replied:
yep good call for now - just I guess eventually it would be good to automatically clean up the PNGs after rendering. Is there a use case for keeping them?

kirek007 replied:
I'm using only PNG files actually! I never render osd directly on video 🤷 
So there should be a switch maybe just to render png? and stop. Then another option to render video and remove png. That makes a lot of sense to me.

odgrace replied:
maybe just adding a 

--cleanup-pngs (default=False) help='removes PNGs after rendering video'. This would only work if you're actually rendering video.

See below about what the API looks like but yes you could also add a --render-video flag if you wanted ffmpeg to do the overlay on the video. As you can see with what I've done I've implicitly assumed everyone just wants ffmpeg to do it!

kirek007 commented on cli.py:
```diff
@@ -0,0 +1,72 @@
+
+    args = parser.parse_args()
+
+    def implicit_path(args, ext):
```
I'll have to move this function to utils. But it can stay like this for now

kirek007 reviewed (commented):
What I'm missing here is the possibility to render one test frame so the user can validate provided parameters.
Another thing would be the creation of requirements-noui.txt files so it should work on any platform without crazy hacking around wxPython.

This week I was super busy, but it should be better next week. Then I can improve it.

odgrace commented:
> What I'm missing here is the possibility to render one test frame so the user can validate provided parameters. 

We could probably split this out so its like 
```python3 cli.py test-frame --args```

compared with something like the below for the main function
```python3 cli.py render-video --args```

I've never done this before but I'm certain it's possible with argparse.

> Another thing would be the creation of requirements-noui.txt files so it should work on any platform without crazy hacking around wxPython.

Yes good point - like how does this get packaged up. I got this working by cloning and just commenting out wxPython requirement because I had the issue with attrdict and I never really wanted the CLI! I don't know how this would work - like if this were on pypi would you have two packages, one derives from another? Like a core/CLI and -gui version? GUI takes core as a requirement and adds wx etc.

kirek007 commented:
The CLI should be fairly simple to package into executables for all platforms. I would need to test it first tho.

odgrace committed:
Added output_path parameter to cli as location to store the PNGs and finished video.

Follows discussion on https://github.com/kirek007/ws-osd-py/pull/43. Default parameter is set to current directory if not specified.

odgrace committed:
Moved function out of main block for readability. Will move to Utils later.

odgrace committed:
Added flag for not rendering video. Useful if you only want the PNGs for other software

odgrace committed:
Amended output path default to make a fresh directory. This avoids dumping 4,000 PNGs in your current folder...

odgrace committed:
Added flag for removing PNG folder once the video has been rendered. This is placed into a try/finally block to remove if flagged even if some error
happens in the render pipeline.

kirek007 committed:
chore(readme): Add basic cli doc

kirek007 committed:
feat(cli): Enable HW acceleration by default

kirek007 committed:
chore(deps): Add dedicated noUI requirements

kirek007 commented:
Hey @odgrace It took some time, but I've reviewed it, added a few tweaks and for me, it's ready to merge :)

odgrace reviewed (commented):
small point to mention, --use-hw now returns true, whether or not you specify it! So maybe we need to switch this to --no-hw and store_false, default=False.

Side note: with this default on my Mac M1, this correctly selects the fast Mac h264 encoder BUT that encode has a stupid quality default (1/100...)
